apps/omg_watcher/lib/omg_watcher/block_validator.ex

```diff
@@ -64,4 +73,93 @@ defmodule OMG.Watcher.BlockValidator do
       end
     end)
   end
+
+  @spec number_of_transactions_within_limit([Transaction.Signed.tx_bytes()]) ::
+          :ok | {:error, atom()}
+  defp number_of_transactions_within_limit(transactions) do
+    case length(transactions) do
+      # A block with a merge transaction only would have no fee transaction.
+      0 ->
+        {:error, :empty_block}
+
+      n when n > @transaction_upper_limit ->
+        {:error, :transactions_exceed_block_limit}
+
+      _ ->
+        :ok
+    end
+  end
+
+  @spec verify_no_duplicate_inputs([Transaction.Recovered.t()]) :: {:ok, map()}
+  defp verify_no_duplicate_inputs(transactions) do
+    Enum.reduce_while(transactions, {:ok, %{}}, fn tx, {:ok, input_set} ->
```

Comment:
this looks like a reduce abuse + scan_for_duplicates. A simple recursive function will be more readable.

Reply:
or not, but this doesn't look too clean!

Reply:
Ino hates `reduce_while` :trollface:

Reply:
yeah when I see things like `{:halt, {:halt, {:error, :block_duplicate_inputs}}}` 🗡️ 🗡️ 🗡️  :)))))

Reply:
Oh I missed the double `:halt`, needs some refactor indeed!

Reply:
Haha point taken. I was going for performance but it got a bit ugly. Refactored this.